### ISSUE TITLE
[R] Unable to install arrow on Fedora40 undefined symbol: _ZTIN6apache6thrift8protocol9TProtocolE

### ISSUE THREAD

**Issue by gowerc** [apache/arrow#43337]
### Describe the bug, including details regarding any error messages, version, and platform.

Heya,

Am currently trying to install the latest R CRAN version of arrow but keep getting the following error:

```
> install.packages("arrow")
Installing package into ‘/home/gowerc/R/x86_64-redhat-linux-gnu-library/4.4’
(as ‘lib’ is unspecified)
trying URL 'https://www.stats.bris.ac.uk/R/src/contrib/arrow_16.1.0.tar.gz'
Content type 'application/x-gzip' length 4439675 bytes (4.2 MB)
==================================================
downloaded 4.2 MB

* installing *source* package ‘arrow’ ...
** package ‘arrow’ successfully unpacked and MD5 sums checked
** using staged installation
*** pkg-config found.
*** Trying Arrow C++ found by pkg-config: /usr
**** Not using: C++ library version (15.0.2): not supported by R package version 16.1.0
*** Found local C++ source: 'tools/cpp'
*** Building libarrow from source
    For build options and troubleshooting, see the install guide:
    https://arrow.apache.org/docs/r/articles/install.html
**** cmake 3.28.2: /usr/bin/cmake
**** arrow 
PKG_CFLAGS=-DUTF8PROC_EXPORTS -I/tmp/RtmpfZTpu0/R.INSTALL37cd0e2a5202/arrow/libarrow/arrow-16.1.0/include -DARROW_STATIC -DPARQUET_STATIC   -DARROW_R_WITH_PARQUET -DARROW_R_WITH_DATASET -DARROW_R_WITH_ACERO -DARROW_R_WITH_JSON
PKG_LIBS=-L/tmp/RtmpfZTpu0/R.INSTALL37cd0e2a5202/arrow/libarrow/arrow-16.1.0/lib -lsnappy -llz4 -lutf8proc -larrow_dataset -larrow -larrow_bundled_dependencies -ldl -lparquet -larrow_acero 
** libs
using C++ compiler: ‘g++ (GCC) 14.1.1 20240701 (Red Hat 14.1.1-7)’
using C++17
g++ -std=gnu++17 -I"/usr/include/R" -DNDEBUG -DUTF8PROC_EXPORTS -I/tmp/RtmpfZTpu0/R.INSTALL37cd0e2a5202/arrow/libarrow/arrow-16.1.0/include -DARROW_STATIC -DPARQUET_STATIC   -DARROW_R_WITH_PARQUET -DARROW_R_WITH_DATASET -DARROW_R_WITH_ACERO -DARROW_R_WITH_JSON -I'/home/gowerc/R/x86_64-redhat-linux-gnu-library/4.4/cpp11/include' -I/usr/local/include    -fpic  -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64 -march=x86-64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer    -c RTasks.cpp -o RTasks.o


<lots of compiler log skipped>


** building package indices
** testing if installed package can be loaded from temporary location
Error: package or namespace load failed for ‘arrow’ in dyn.load(file, DLLpath = DLLpath, ...):
 unable to load shared object '/home/gowerc/R/x86_64-redhat-linux-gnu-library/4.4/00LOCK-arrow/00new/arrow/libs/arrow.so':
  /home/gowerc/R/x86_64-redhat-linux-gnu-library/4.4/00LOCK-arrow/00new/arrow/libs/arrow.so: undefined symbol: _ZTIN6apache6thrift8protocol9TProtocolE
Error: loading failed
Execution halted
ERROR: loading failed
* removing ‘/home/gowerc/R/x86_64-redhat-linux-gnu-library/4.4/arrow’

The downloaded source packages are in
        ‘/tmp/RtmpiE3fZa/downloaded_packages’
Warning message:
In install.packages("arrow") :
  installation of package ‘arrow’ had non-zero exit status
```

## General Debug Information

```
> version
               _                           
platform       x86_64-redhat-linux-gnu     
arch           x86_64                      
os             linux-gnu                   
system         x86_64, linux-gnu           
status                                     
major          4                           
minor          4.1                         
year           2024                        
month          06                          
day            14                          
svn rev        86737                       
language       R                           
version.string R version 4.4.1 (2024-06-14)
nickname       Race for Your Life         
```

```
> cat /etc/os-release
NAME="Fedora Linux"
VERSION="40 (Workstation Edition)"
ID=fedora
VERSION_ID=40
VERSION_CODENAME=""
PLATFORM_ID="platform:f40"
PRETTY_NAME="Fedora Linux 40 (Workstation Edition)"
ANSI_COLOR="0;38;2;60;110;180"
LOGO=fedora-logo-icon
CPE_NAME="cpe:/o:fedoraproject:fedora:40"
DEFAULT_HOSTNAME="fedora"
HOME_URL="https://fedoraproject.org/"
DOCUMENTATION_URL="https://docs.fedoraproject.org/en-US/fedora/f40/system-administrators-guide/"
SUPPORT_URL="https://ask.fedoraproject.org/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
REDHAT_BUGZILLA_PRODUCT="Fedora"
REDHAT_BUGZILLA_PRODUCT_VERSION=40
REDHAT_SUPPORT_PRODUCT="Fedora"
REDHAT_SUPPORT_PRODUCT_VERSION=40
SUPPORT_END=2025-05-13
VARIANT="Workstation Edition"
VARIANT_ID=workstation
```

Not sure if relevant but I had installed the parquet dev lib from `dnf` for separate project:
```
> sudo dnf info parquet-libs-devel                                 
Last metadata expiration check: 2:39:56 ago on Fri 19 Jul 2024 13:27:15 BST.
Installed Packages
Name         : parquet-libs-devel
Version      : 15.0.2
Release      : 3.fc40
Architecture : x86_64
Size         : 489 k
Source       : libarrow-15.0.2-3.fc40.src.rpm
Repository   : @System
From repo    : fedora
Summary      : Libraries and header files for Apache Parquet C++
URL          : https://arrow.apache.org/
License      : Apache-2.0
Description  : Libraries and header files for Apache Parquet C++.
```

### Component(s)

R

**Comment by assignUser** [comment]
@gowerc could you upload the full cmake log as a gist or pastebin?

**Comment by assignUser** [comment]
>  (if that helps at all...)

Yes very good thank you!

**Comment by assignUser** [comment]
Could you re-run it with `ARROW_R_DEV=true` as an envvar so we get the full cmake log?

As workaround to unblock you can install with `LIBARROW_BINARY=true` that should fetch a pre-compiled binary matching your system.

**Comment by assignUser** [comment]
Hm, I think this is the issue: https://gist.github.com/gowerc/6c96fb454fb3f002daece58888fdfe78#file-log-txt-L463
I guess the .so file gets lost after the package is build and installed... we probably need to enforce `ARROW_DEPENDENCY_USE_SHARED` in the build script.

We are currently working on the next CRAN release, I will investigate and make sure a patch is included cc @jonkeane 

@gowerc did installing with the precompiled binary work?

**Comment by assignUser** [comment]
Trying to repro this in docker now to get a fix going so we can include it in the CRAN release of 17.0.0

**Comment by assignUser** [comment]
Ok I think the issue is different than I thought and also happens with the above option set... maybe it is caused by gcc 14 on fedora? It doesn't seem to be a gcc 14 issue as I was able to build on ubuntu with 14 and on fedora 38 with gcc13.

I am not familiar enough with fedora to really troubleshoot this further at the moment, sorry. Though I have confirmed that `LIBARROW_BINARY=true` works, you only need to have `libcurl-devel openssl-devel` installed.

**Comment by kou** [comment]
Hmm. It seems that this is caused by wrong link order:

https://gist.github.com/gowerc/6c96fb454fb3f002daece58888fdfe78#file-log-txt-L1499

```text
g++ -std=gnu++17 -shared -L/usr/lib64/R/lib -Wl,-z,relro -Wl,--as-needed -Wl,-z,pack-relative-relocs -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld-errors -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -Wl,--build-id=sha1 -o arrow.so RTasks.o altrep.o array.o array_to_vector.o arraydata.o arrowExports.o bridge.o buffer.o chunkedarray.o compression.o compute-exec.o compute.o config.o csv.o dataset.o datatype.o expression.o extension-impl.o feather.o field.o filesystem.o io.o json.o memorypool.o message.o parquet.o r_to_arrow.o recordbatch.o recordbatchreader.o recordbatchwriter.o safe-call-into-r-impl.o scalar.o schema.o symbols.o table.o threadpool.o type_infer.o -L/tmp/RtmpQmOHjc/R.INSTALL46f6c3a31d910/arrow/libarrow/arrow-16.1.0/lib -lsnappy -llz4 -lutf8proc -larrow_dataset -larrow -larrow_bundled_dependencies -ldl -lparquet -larrow_acero -L/usr/lib64/R/lib -lR
```

Here is the `-l` options part in the command line:

```text
-lsnappy
-llz4
-lutf8proc
-larrow_dataset
-larrow
-larrow_bundled_dependencies
-ldl
-lparquet
-larrow_acero
-L/usr/lib64/R/lib
-lR
```

It should be:

```text
-larrow_dataset
-lparquet
-larrow_acero
-larrow
-larrow_bundled_dependencies
-ldl
-lsnappy
-llz4
-lutf8proc
-L/usr/lib64/R/lib
-lR
```

**Comment by kou** [comment]
Can we share `arrow*.pc` and `parquet.pc`?
For example, https://gist.github.com/gowerc/6c96fb454fb3f002daece58888fdfe78#file-log-txt-L1001 is the location of `arrow.pc`. Other `*.pc` should exist in the same directory.

BTW, it's not related to this problem but it seems link order without `pkg-config` is wrong:

```diff
diff --git a/r/configure b/r/configure
index 0882ee6719..1bad59ad9d 100755
--- a/r/configure
+++ b/r/configure
@@ -347,13 +347,6 @@ add_feature_flags () {
       # NOTE: parquet is assumed to have the same -L flag as arrow
       # so there is no need to add its location to PKG_DIRS
     fi
-    if arrow_built_with ARROW_DATASET; then
-      PKG_CFLAGS_FEATURES="$PKG_CFLAGS_FEATURES -DARROW_R_WITH_DATASET"
-      PKG_CONFIG_NAMES_FEATURES="$PKG_CONFIG_NAMES_FEATURES arrow-dataset"
-      PKG_LIBS_FEATURES_WITHOUT_PC="-larrow_dataset $PKG_LIBS_FEATURES_WITHOUT_PC"
-      # NOTE: arrow_dataset is assumed to have the same -L flag as arrow
-      # so there is no need to add its location to PKG_DIRS
-    fi
     if arrow_built_with ARROW_ACERO; then
       PKG_CFLAGS_FEATURES="$PKG_CFLAGS_FEATURES -DARROW_R_WITH_ACERO"
       PKG_CONFIG_NAMES_FEATURES="$PKG_CONFIG_NAMES_FEATURES arrow-acero"
@@ -361,6 +354,13 @@ add_feature_flags () {
       # NOTE: arrow_acero is assumed to have the same -L flag as arrow
       # so there is no need to add its location to PKG_DIRS
     fi
+    if arrow_built_with ARROW_DATASET; then
+      PKG_CFLAGS_FEATURES="$PKG_CFLAGS_FEATURES -DARROW_R_WITH_DATASET"
+      PKG_CONFIG_NAMES_FEATURES="$PKG_CONFIG_NAMES_FEATURES arrow-dataset"
+      PKG_LIBS_FEATURES_WITHOUT_PC="-larrow_dataset $PKG_LIBS_FEATURES_WITHOUT_PC"
+      # NOTE: arrow_dataset is assumed to have the same -L flag as arrow
+      # so there is no need to add its location to PKG_DIRS
+    fi
     if arrow_built_with ARROW_SUBSTRAIT; then
       PKG_CFLAGS_FEATURES="$PKG_CFLAGS_FEATURES -DARROW_R_WITH_SUBSTRAIT"
       PKG_CONFIG_NAMES_FEATURES="$PKG_CONFIG_NAMES_FEATURES arrow-substrait"
```

`libarrow_dataset` depends on `libarrow_acero`. So `-larrow_dataset -larrow_acero` order is correct.

**Comment by assignUser** [comment]
Thanks @kou here are the files:
[pcfiles.tar.gz](https://github.com/user-attachments/files/16325658/pcfiles.tar.gz)

(wow I just saw today, monday, that apparently my comment was never send)

**Comment by kou** [comment]
Thanks.

It seems that `pkgconf` in Fedora 40 has a problem:

```console
$ docker run -it --rm -v $PWD:/host fedora:40
$ dnf install -y pkgconf
$ rpm -qa | grep pkgconf
pkgconf-m4-2.1.1-1.fc40.noarch
libpkgconf-2.1.1-1.fc40.x86_64
pkgconf-2.1.1-1.fc40.x86_64
pkgconf-pkg-config-2.1.1-1.fc40.x86_64
$ PKG_CONFIG_PATH=/host/pcfiles pkgconf --libs arrow-dataset parquet
-L/tmp/Rtmpo6n7mE/R.INSTALL10001961ec66/arrow/libarrow/arrow-16.1.0/lib -larrow_dataset -larrow_acero -larrow -larrow_bundled_dependencies -ldl -lparquet
````

`-lparquet` should exist before `-larrow` but it exists at the end.

Here is a result on Debian GNU/Linux sid:

```bash
$ dpkg -l | grep pkgconf
ii  libextutils-pkgconfig-perl                               1.16-3                                    all          Perl interface to the pkg-config utility
ii  libpkgconf3:amd64                                        1.8.1-3                                   amd64        shared library for pkgconf
ii  pkgconf:amd64                                            1.8.1-3                                   amd64        manage compile and link flags for libraries
ii  pkgconf-bin                                              1.8.1-3                                   amd64        manage compile and link flags for libraries (binaries)
ii  python3-pkgconfig                                        1.5.5-2                                   all          Python interface to the pkg-config command line tool
$ PKG_CONFIG_PATH=/tmp/download/pcfiles pkgconf --libs arrow-dataset parquet
-L/tmp/Rtmpo6n7mE/R.INSTALL10001961ec66/arrow/libarrow/arrow-16.1.0/lib -larrow_dataset -larrow_acero -lparquet -larrow -larrow_bundled_dependencies -ldl
```

`-lparquet` exists before `-larrow`.

I also tried https://github.com/pkgconf/pkgconf master:

```bash
$ git clone git@github.com:pkgconf/pkgconf.git
$ meson setup --prefix=/tmp/local pkgconf.build pkgconf
$ ninja -C pkgconf.build
$ pkgconf.build/pkgconf --version
2.2.0
$ PKG_CONFIG_PATH=/tmp/download/pcfiles pkgconf.build/pkgconf --libs arrow-dataset parquet
-L/tmp/Rtmpo6n7mE/R.INSTALL10001961ec66/arrow/libarrow/arrow-16.1.0/lib -larrow_dataset -larrow_acero -lparquet -larrow -larrow_bundled_dependencies -ldl
```

`-lparquet` exists before `-larrow`.

It seems that `pkgconf` in Fedora 40 has a problem. https://bugzilla.redhat.com/show_bug.cgi?id=2298067 may be related.

**Comment by assignUser** [comment]
Thank you for investigating and the informative code blocks @kou! 

I am going to close this issue as it's not something we can fix on our end, let's hope they upgrade the version in the fedora repo to 2.2+ soon.

@gowerc You already have a workaround but if you want to build from source for some reason you can disable the use of pc via the envvar `ARROW_USE_PKG_CONFIG=false` though with the bug kou pointed out above it might also error, I will see that the patch is included in 17.0.0.

**Comment by kou** [comment]
If you're interesting in this, how about opening a pull request like https://src.fedoraproject.org/rpms/pkgconf/pull-request/10#request_diff ?

**Comment by david-cortes** [comment]
@assignUser I get the same error on debian 12 with the arrow that gets compiled on-the-fly when calling `install.packages` without any existing `libarrow` installation. The error happens whether I use `gcc` or `clang` as compiler. Albeit with `clang` it complains about a different symbol `_ZN6apache6thrift8protocol16TProtocolFactoryD2Ev`.

**Comment by kou** [comment]
Hmm. Debian 12 doesn't use pkgconf 2.1.1: https://packages.debian.org/bookworm/pkgconf

It may be a different issue. Could you open a new issue and add your full log to it like this issue's description?
